doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1748,31 +1757,39 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.  If
+   :n:`@one_term_with_bindings` is not provided, the tactic first does a
+   :tacn:`intros`
+   then succeeds if there is a hypothesis :n:`False` in the updated context, or
+   if there is a pair of hypotheses where one is simple negation of the other
+   (i.e. :n:`P` and not :n:`~P`).
 
-   .. exn:: No such assumption.
-      :undocumented:
+   Use the :tacn:`discriminate` tactic to prove the current goal when there
+   is a hypothesis with an impossible structural equality such as
+   :n:`0 = 1`.
 
-.. tacv:: contradiction @ident
+   If :n:`@one_term_with_bindings` has the type :n:`False`, the tactic
+   proves the current goal.  If the type is a negation, e.g. :n:`~P`, then
```

Comment:
It looks like the tactic does some reductions (e.g. a cbv?).  Maybe it reduces all the hypotheses before checking for contradictions?  What wording would you suggest?

```
Goal forall (A : Type) (x : A), ~(x = x) -> False.
intros.
(*
1 goal
A : Type@{Top.1}
x : A
H : not (@eq A x x)
______________________________________(1/1)
False
*)
contradiction.
Qed.
```

Reply:
I have no idea what kind of reduction it does. I think the current description is good enough for now (barring my other comments).